tests/functional/test_s3.py

```diff
@@ -1431,6 +1438,55 @@ def test_content_sha256_not_set_if_config_value_is_false(self):
         sha_header = sent_headers.get("x-amz-content-sha256")
         self.assertEqual(sha_header, b"UNSIGNED-PAYLOAD")
 
+    def test_content_sha256_set_if_config_value_not_set_put_object(self):
+        # The default behavior matches payload_signing_enabled=False. For
+        # operations where the `Content-MD5` is present this means that
+        # `x-amz-content-sha256` is present but not set.
+        config = Config(signature_version="s3v4")
+        self.client = self.session.create_client(
+            "s3", self.region, config=config
+        )
+        self.http_stubber = ClientHTTPStubber(self.client)
+        self.http_stubber.add_response()
+        with self.http_stubber:
+            self.client.put_object(Bucket="foo", Key="bar", Body="baz")
+        sent_headers = self.get_sent_headers()
+        sha_header = sent_headers.get("x-amz-content-sha256")
+        self.assertEqual(sha_header, b"UNSIGNED-PAYLOAD")
+
+    def test_content_sha256_set_if_config_value_not_set_list_objects(self):
+        # The default behavior matches payload_signing_enabled=False. For
+        # operations where the `Content-MD5` is not present, this means that
+        # `x-amz-content-sha256` is present and set.
+        config = Config(signature_version="s3v4")
+        self.client = self.session.create_client(
+            "s3", self.region, config=config
+        )
+        self.http_stubber = ClientHTTPStubber(self.client)
+        self.http_stubber.add_response()
+        with self.http_stubber:
+            self.client.list_objects(Bucket="foo")
+        sent_headers = self.get_sent_headers()
+        sha_header = sent_headers.get("x-amz-content-sha256")
+        self.assertNotEqual(sha_header, b"UNSIGNED-PAYLOAD")
```

Comment:
Do we want to also test this is not None? I don't think we're actually proving the header was set.